From 58dcacd193ca34a007f308549b62f9dc7a11fc13 Mon Sep 17 00:00:00 2001
From: Timon Sassor <timon.sassor@governikus.de>
Date: Fri, 5 Sep 2025 11:38:43 +0200
Subject: a11y Android: Enable granular text navigation for editable components

Talkback allows it to change the text navigation on editable components
between character, word, line, sentence etc.. However the Android
platform plugin does not listen to these gestures and does not pass them
to editable components like TextEdit.
So we listen to these actions and provide the appropriate events.
QAccessibleTextInterface already has the ability to navigate text as
described. So we add the missing link here, which is the support in the
accessibility plugin.

[ChangeLog][Accessibility][Android] Enable granular text navigation for
editable components e.g. TextEdit

Fixes: QTBUG-139943
Change-Id: I118222bc23c192d3bc9198f639166f39801691e8
---
 .../qt/android/QtAccessibilityDelegate.java   |  34 ++++-
 .../qt/android/QtNativeAccessibility.java     |   3 +
 .../android/androidjniaccessibility.cpp       | 117 ++++++++++++++++++
 3 files changed, 152 insertions(+), 2 deletions(-)

diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
index db63acd4153376fe965d655a6aecb60b90daa8cf..81db2da44d866e12c454af04f5e5c1d16ddf6f9b 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
@@ -43,6 +43,8 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
     // When exploring the screen by touch, the item "hovered" by the finger.
     private int m_hoveredVirtualViewId = INVALID_ID;
 
+    private int m_currentRotorSetting = AccessibilityNodeInfo.MOVEMENT_GRANULARITY_CHARACTER;
+
     // Cache coordinates of the view to know the global offset
     // this is because the Android platform window does not take
     // the offset of the view on screen into account (eg status bar on top)
@@ -382,6 +384,12 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
 
         event.setPackageName(m_view.getContext().getPackageName());
         event.setSource(m_view, virtualViewId);
+
+        if (eventType == AccessibilityEvent.TYPE_VIEW_TEXT_TRAVERSED_AT_MOVEMENT_GRANULARITY) {
+            event.setMovementGranularity(m_currentRotorSetting);
+            QtNativeAccessibility.populateGranularMovementEvent(virtualViewId, event);
+        }
+
         return event;
     }
 
@@ -497,6 +505,17 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
         screenRect.offset(-parentScreenRect.left, -parentScreenRect.top);
         node.setBoundsInParent(screenRect);
 
+        if (node.isEnabled() && node.isEditable()) {
+            node.addAction(AccessibilityNodeInfo.AccessibilityAction.ACTION_NEXT_AT_MOVEMENT_GRANULARITY);
+            node.addAction(AccessibilityNodeInfo.AccessibilityAction.ACTION_PREVIOUS_AT_MOVEMENT_GRANULARITY);
+            node.setMovementGranularities(
+                AccessibilityNodeInfo.MOVEMENT_GRANULARITY_CHARACTER |
+                AccessibilityNodeInfo.MOVEMENT_GRANULARITY_WORD |
+                AccessibilityNodeInfo.MOVEMENT_GRANULARITY_LINE |
+                AccessibilityNodeInfo.MOVEMENT_GRANULARITY_PARAGRAPH
+            );
+        }
+
         // Manage internal accessibility focus state.
         if (m_focusedVirtualViewId == virtualViewId) {
             node.setAccessibilityFocused(true);
@@ -561,13 +580,13 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
                         return m_view.performAccessibilityAction(action, arguments);
                     }
             }
-            handled |= performActionForVirtualViewId(virtualViewId, action);
+            handled |= performActionForVirtualViewId(virtualViewId, action, arguments);
 
             return handled;
         }
     };
 
-    protected boolean performActionForVirtualViewId(int virtualViewId, int action)
+    protected boolean performActionForVirtualViewId(int virtualViewId, int action, Bundle arguments)
     {
         //noinspection CommentedOutCode
         {
@@ -596,6 +615,17 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
         case AccessibilityNodeInfo.ACTION_SCROLL_BACKWARD:
             success = QtNativeAccessibility.scrollBackward(virtualViewId);
             break;
+        case AccessibilityNodeInfo.ACTION_NEXT_AT_MOVEMENT_GRANULARITY:
+        case AccessibilityNodeInfo.ACTION_PREVIOUS_AT_MOVEMENT_GRANULARITY:
+            m_currentRotorSetting = arguments.getInt(
+                AccessibilityNodeInfo.ACTION_ARGUMENT_MOVEMENT_GRANULARITY_INT,
+                AccessibilityNodeInfo.MOVEMENT_GRANULARITY_CHARACTER // Default-Value, if not set
+            );
+            final boolean movingForward = action == AccessibilityNodeInfo.ACTION_NEXT_AT_MOVEMENT_GRANULARITY;
+            success = QtNativeAccessibility.moveToNextElementAtGranularity(virtualViewId, movingForward, m_currentRotorSetting);
+            if (success)
+                sendEventForVirtualViewId(virtualViewId, AccessibilityEvent.TYPE_VIEW_TEXT_TRAVERSED_AT_MOVEMENT_GRANULARITY);
+            break;
         }
         return success;
     }
diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
index 36898fc3a840b16cca35918f7e76dbe3455510d0..9074e589a779c0bf20c383f0cb0fd40642d49111 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
@@ -6,6 +6,7 @@ package org.qtproject.qt.android;
 import android.graphics.Rect;
 import android.view.accessibility.AccessibilityEvent;
 import android.view.accessibility.AccessibilityNodeInfo;
+import android.view.accessibility.AccessibilityEvent;
 
 class QtNativeAccessibility
 {
@@ -20,7 +21,9 @@ class QtNativeAccessibility
     static native boolean focusAction(int objectId);
     static native boolean scrollForward(int objectId);
     static native boolean scrollBackward(int objectId);
+    static native boolean moveToNextElementAtGranularity(int objectId, boolean forward, int granularity);
 
     static native boolean populateNode(int objectId, AccessibilityNodeInfo node);
     static native boolean populateScrollEvent(int objectId, AccessibilityEvent event);
+    static native boolean populateGranularMovementEvent(int objectId, AccessibilityEvent event);
 }
diff --git x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
index a6e587291d79b75cc17bfed4a8f5803b6c132022..d1611aef02fc24f7b6fd38a4e29b2a2fa0079aaa 100644
--- x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
@@ -52,6 +52,17 @@ namespace QtAndroidAccessibility
     static jmethodID m_setToIndexMethodID = 0;
     static jmethodID m_setItemCountMethodID = 0;
 
+    static jmethodID m_getMovementGranularityEventMethodID = 0;
+    static jmethodID m_setFromIndexEventMethodID = 0;
+    static jmethodID m_setToIndexEventMethodID = 0;
+
+    static const QMap<QAccessible::TextBoundaryType, int> m_movementGranularityMapping {
+        std::make_pair(QAccessible::TextBoundaryType::CharBoundary, QJniObject::getStaticField<int>("android/view/accessibility/AccessibilityNodeInfo", "MOVEMENT_GRANULARITY_CHARACTER")),
+        std::make_pair(QAccessible::TextBoundaryType::WordBoundary, QJniObject::getStaticField<int>("android/view/accessibility/AccessibilityNodeInfo", "MOVEMENT_GRANULARITY_WORD")),
+        std::make_pair(QAccessible::TextBoundaryType::LineBoundary, QJniObject::getStaticField<int>("android/view/accessibility/AccessibilityNodeInfo", "MOVEMENT_GRANULARITY_LINE")),
+        std::make_pair(QAccessible::TextBoundaryType::ParagraphBoundary, QJniObject::getStaticField<int>("android/view/accessibility/AccessibilityNodeInfo", "MOVEMENT_GRANULARITY_PARAGRAPH"))
+    };
+
     static bool m_accessibilityActivated = false;
 
     // This object is needed to schedule the execution of the code that
@@ -413,6 +424,50 @@ namespace QtAndroidAccessibility
         return result && oldPosition != screenRect_helper(firstChildId, false);
     }
 
+    static bool moveToNextElementAtGranularity_helper(int objectId, bool forward, int granularity)
+    {
+        QAccessibleInterface *iface = interfaceFromId(objectId);
+        if (!iface || !iface->isValid() || !iface->textInterface())
+            return false;
+
+        QAccessibleTextInterface *textIface = iface->textInterface();
+
+        Q_ASSERT(m_movementGranularityMapping.first() != 0);
+        const auto movementGranularity = m_movementGranularityMapping.key(granularity, QAccessible::TextBoundaryType::NoBoundary);
+        const int cursorPosition = textIface->cursorPosition();
+        const int charCount = textIface->characterCount();
+        const bool wouldMoveOutsideBounds = (!forward && cursorPosition == 0) || (forward && cursorPosition == charCount);
+        if (movementGranularity == QAccessible::TextBoundaryType::NoBoundary || wouldMoveOutsideBounds)
+            return false;
+
+        int fromIndex = 0;
+        int toIndex = 0;
+        if (forward)
+            textIface->textAfterOffset(cursorPosition, movementGranularity, &fromIndex, &toIndex);
+        else
+            textIface->textBeforeOffset(cursorPosition, movementGranularity, &fromIndex, &toIndex);
+
+        if (toIndex < 0)
+            textIface->setCursorPosition(charCount);
+        else if (fromIndex < 0)
+            textIface->setCursorPosition(0);
+        else
+            textIface->setCursorPosition(fromIndex);
+
+        return true;
+    }
+
+    static jboolean moveToNextElementAtGranularity(JNIEnv */*env*/, jobject /*thiz*/, jint objectId, jboolean forward, jint granularity)
+    {
+        bool result = false;
+        if (m_accessibilityContext) {
+            runInObjectContext(m_accessibilityContext, [objectId, forward, granularity]() {
+                return moveToNextElementAtGranularity_helper(objectId, forward, granularity);
+            }, &result);
+        }
+        return result;
+    }
+
     static QString textFromValue(QAccessibleInterface *iface)
     {
         QString valueStr;
@@ -863,6 +918,61 @@ namespace QtAndroidAccessibility
         return false;
     }
 
+    struct GranularMovementEventInfo
+    {
+        bool valid = false;
+        int fromIndex = 0;
+        int toIndex = 0;
+    };
+
+    static GranularMovementEventInfo populateGranularMovementEvent_helper(int objectId, int movementGranularity)
+    {
+        GranularMovementEventInfo info;
+        QAccessibleInterface *iface = interfaceFromId(objectId);
+        if (!iface || !iface->isValid())
+            return info;
+
+        info.valid = true;
+
+        QAccessibleTextInterface *textIface = iface->textInterface();
+        if (!textIface)
+            return info;
+
+        const auto mappedMovementGranularity = m_movementGranularityMapping.key(movementGranularity, QAccessible::TextBoundaryType::NoBoundary);
+        if (mappedMovementGranularity != QAccessible::TextBoundaryType::NoBoundary) {
+            // Since this method is called after the cursor already moved, we only have to look left of the cursor.
+            textIface->textBeforeOffset(
+                textIface->cursorPosition(),
+                mappedMovementGranularity,
+                &info.fromIndex,
+                &info.toIndex
+            );
+        }
+
+        return info;
+    }
+
+    static jboolean populateGranularMovementEvent(JNIEnv *env, jobject /*thiz*/, jint objectId, jobject event)
+    {
+        GranularMovementEventInfo info;
+        const jint movementGranularity = m_movementGranularityMapping.first();
+        env->CallIntMethod(event, m_getMovementGranularityEventMethodID, &movementGranularity);
+        if (m_accessibilityContext) {
+            runInObjectContext(m_accessibilityContext, [objectId, &movementGranularity]() {
+                return populateGranularMovementEvent_helper(objectId, movementGranularity);
+            }, &info);
+        }
+        if (!info.valid) {
+            __android_log_print(ANDROID_LOG_WARN, m_qtTag, "Accessibility: populateGranularMovementEvent for Invalid ID");
+            return false;
+        }
+
+        env->CallVoidMethod(event, m_setFromIndexEventMethodID, info.fromIndex);
+        env->CallVoidMethod(event, m_setToIndexEventMethodID, info.toIndex);
+
+        return true;
+    }
+
     static const JNINativeMethod methods[] = {
         {"setActive","(Z)V",(void*)setActive},
         {"childIdListForAccessibleObject", "(I)[I", (jintArray)childIdListForAccessibleObject},
@@ -877,6 +987,8 @@ namespace QtAndroidAccessibility
         {"focusAction", "(I)Z", (void*)focusAction},
         {"scrollForward", "(I)Z", (void*)scrollForward},
         {"scrollBackward", "(I)Z", (void*)scrollBackward},
+        {"moveToNextElementAtGranularity", "(IZI)Z", (void *)moveToNextElementAtGranularity},
+        {"populateGranularMovementEvent", "(ILandroid/view/accessibility/AccessibilityEvent;)Z", (void *)populateGranularMovementEvent}
     };
 
 #define GET_AND_CHECK_STATIC_METHOD(VAR, CLASS, METHOD_NAME, METHOD_SIGNATURE) \
@@ -924,6 +1036,11 @@ namespace QtAndroidAccessibility
         GET_AND_CHECK_STATIC_METHOD(m_setToIndexMethodID, eventClass, "setToIndex", "(I)V");
         GET_AND_CHECK_STATIC_METHOD(m_setItemCountMethodID, eventClass, "setItemCount", "(I)V");
 
+        jclass accessibilityEventClass = env->FindClass("android/view/accessibility/AccessibilityEvent");
+        GET_AND_CHECK_STATIC_METHOD(m_getMovementGranularityEventMethodID, accessibilityEventClass, "getMovementGranularity", "()I");
+        GET_AND_CHECK_STATIC_METHOD(m_setFromIndexEventMethodID, accessibilityEventClass, "setFromIndex", "(I)V");
+        GET_AND_CHECK_STATIC_METHOD(m_setToIndexEventMethodID, accessibilityEventClass, "setToIndex", "(I)V");
+
         return true;
     }
 }
